tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -245,12 +200,114 @@ tfw_sched_hash_get_srv_conn(TfwMsg *msg, TfwServer *srv)
 	return NULL;
 }
 
+static void
+tfw_sched_hash_cleanup(TfwSrvGroup *sg)
+{
+	size_t si;
+	TfwHashSrvList *sl = sg->sched_data;
+
+	if (!sl)
+		return;
+
+	for (si = 0; si < sl->srv_n; ++si) {
+		if (sl->srvs[si].conn)
+			kfree(sl->srvs[si].conn);
+		if (sl->srvs[si].hash)
+			kfree(sl->srvs[si].hash);
+	}
+
+	kfree(sl);
+	sg->sched_data = NULL;
+}
+
+static void
+tfw_sched_hash_del_grp(TfwSrvGroup *sg)
+{
+	tfw_sched_hash_cleanup(sg);
+}
+
+/**
+ * Validate the integrity of a group.
+ *
+ * Make sure that number of servers in the group, and the number
+ * of connections for each server match the recorded values.
+ */
+static int
+tfw_sched_hash_validate_grp(TfwSrvGroup *sg)
+{
+	size_t si = 0, ci;
+	TfwServer *srv;
+	TfwSrvConn *srv_conn;
+
+	list_for_each_entry(srv, &sg->srv_list, list) {
+		ci = 0;
+		list_for_each_entry(srv_conn, &srv->conn_list, list)
+			++ci;
+		if (ci > srv->conn_n)
+			return -EINVAL;
+		++si;
+	}
+	if (si > sg->srv_n)
+		return -EINVAL;
+
+	return 0;
+}
+
+static int
+tfw_sched_hash_add_grp(TfwSrvGroup *sg)
+{
+	int ret = -ENOMEM;
+	size_t size, ci;
+	unsigned int sum_conn_n;
+	TfwServer *srv;
+	TfwSrvConn *srv_conn;
+	TfwHashSrv *hsrv;
+	TfwHashSrvList *sl;
+
+	if (tfw_sched_hash_validate_grp(sg))
+		return -EINVAL;
+
+	size = sizeof(TfwHashSrvList) + sizeof(TfwHashSrv) * sg->srv_n;
+	if (!(sg->sched_data = kzalloc(size, GFP_KERNEL)))
+		return -ENOMEM;
+	sl = sg->sched_data;
+	sl->srvs = sg->sched_data + sizeof(TfwHashSrvList);
+	sl->srv_n = sg->srv_n;
+
+	sum_conn_n = 0;
+	hsrv = sl->srvs;
+	list_for_each_entry(srv, &sg->srv_list, list) {
+		size = sizeof(hsrv->conn[0]) * srv->conn_n;
+		if (!(hsrv->conn = kzalloc(size, GFP_KERNEL)))
+			goto cleanup;
+		size = sizeof(hsrv->hash[0]) * srv->conn_n;
+		if (!(hsrv->hash = kzalloc(size, GFP_KERNEL)))
+			goto cleanup;
```

Comment:
Yeah, I missed that we need `srv->conn_n`, so probably it'd be really hard to use only one allocation. However, these two particular calls of `kzalloc()` depend on the same data and must be merged.